### PR TITLE
handle transpose during model loading with TP (Qwen3-VL-MoE)

### LINKED ISSUE #48308: Native TP does not shard qwen3_vl_moe experts (weights fully replicated per rank)
Found while testing native TP on Qwen3-VL-30B-A3B; the sibling arch qwen3_5_moe shards correctly with the same code, which makes the comparison easy. Related but distinct from #48204: that draft adds an expert-parallel path; this issue is about the plain tp_size-only path silently replicating the experts.

## Reproduction

```python
# torchrun --nproc-per-node 4 repro.py
import torch
from transformers import AutoModelForImageTextToText
from transformers.distributed.configuration_utils import DistributedConfig

model = AutoModelForImageTextToText.from_pretrained(
    "Qwen/Qwen3-VL-30B-A3B-Instruct", dtype=torch.bfloat16, distributed_config=DistributedConfig(tp_size=4)
)
print(torch.cuda.memory_allocated() / 1e9)  # expected ~15 GB/rank, observed 60.3 GB/rank
```

## Outcome

The model loads without error but each rank holds 60.3 GB: the full 30B-A3B weights. Expert weights are not sharded (for comparison, `Qwen/Qwen3.6-35B-A3B`, arch `qwen3_5_moe`, shards to
10.5 GB/rank with the same code). Consequences observed:

- any training attempt OOMs immediately (weights + grads alone exceed 80 GB)
- continuous batching on the TP model crashes in the paged cache with `index_copy_(): Destination slice shape: 4 128 ... source slice shape: 1 128`: the cache is sized for the config's 4 KV heads while the attention layers are sharded to 1 KV head per rank, so the two disagree (attention appears sharded while the MoE is not).

Root cause: `Qwen3VLMoeTextConfig.base_model_tp_plan` carries dense-style `layers.*.mlp.gate_proj/up_proj/down_proj` entries, paths that do not exist in the MoE layers (experts live at `layers.*.mlp.experts.*`), so no pattern matches the expert weights and they replicate. The sibling `qwen3_5_moe` plan has the fused-expert entries (`experts.gate_up_proj: packed_colwise`, `experts.down_proj: rowwise`, `experts: moe_tp_experts`), which is why it shards correctly.

Precedent: the sibling arch had the identical bug, reported in #45125 and fixed by #45124 (add `_tp_plan` to the ForConditionalGeneration wrapper). That fix was never mirrored to qwen3_vl_moe, which is exactly why the comparison above splits the way it does. #41570 (uneven memory on Qwen3-VL-30B-A3B multi-GPU) may be an earlier symptom sighting.

## Expected behavior

Either a complete TP plan for `qwen3_vl_moe` (experts sharded like the other MoE archs), or a load-time error when parts of the model have no TP mapping, instead of silent replication.

## System info

transformers main (a353632), torch 2.13.0+cu130, Python 3.12.12, 4-8x NVIDIA H100 80GB, torchrun single node.

### PR DESCRIPTION
**Pull request by 3outeille**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48373) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48373)

# Explanation

Specifying a `tp_plan` for `qwen3-vl-moe` is not enough to fix https://github.com/huggingface/transformers/issues/48308.  A fix is as well needed during the loading process.

The loading order is: `checkpoint weight -> shard according to TP plan -> WeightConverter -> load local weight`

The `tp_plan` for `qwen3_vl_moe` is `gate_up_proj: packed_colwise`which means sharding on dimension 1

However, the Qwen3-VL-MoE checkpoint stores this weight transposed: 

```python
 "qwen3_vl_moe": [
            WeightConverter(
                source_patterns="mlp.experts.gate_up_proj",
                target_patterns="mlp.experts.gate_up_proj",
                operations=[Transpose(1, 2, check_dims=True)],
            ),
```

So basically:

```
Model:      [E, 2I, H]  -> 2I is dimension 1
Checkpoint: [E, H, 2I]  -> 2I is dimension 2
```

Before the fix, the checkpoint would be `[E, H, 2I] -> shard checkpoint dim 1 — wrong dimension -> [E, H/2, 2I] -> transpose -> [E, 2I, H/2]` => incorrect 

We could transpose before the the checkpoint but that involves full materialization. Therefore the loader needs to map model dimensions into checkpoint dimensions

The loader therefore needs to map:

```
model dimension 1 -> checkpoint dimension 2
model dimension 2 -> checkpoint dimension 1
```

this way we get, `[E, H, 2I] -> shard checkpoint dim 2 (mapped) -> [E, H, I] -> transpose -> [E, I, H]` => correct 

same thing for `dow_proj`

# Output

```python
# torchrun --nproc-per-node 4 repro.py
import torch
from transformers import AutoModelForImageTextToText, AutoProcessor
from transformers.distributed.configuration_utils import DistributedConfig

model_id = "Qwen/Qwen3-VL-30B-A3B-Instruct"
model = AutoModelForImageTextToText.from_pretrained(
    model_id, dtype=torch.bfloat16, distributed_config=DistributedConfig(tp_size=4)
)
print(torch.cuda.memory_allocated() / 1e9)  # expected ~15 GB/rank

processor = AutoProcessor.from_pretrained(model_id)
messages = [
    {
        "role": "user",
        "content": [
            {
                "type": "image",
                "url": "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/pipeline-cat-chonk.jpeg",
            },
            {"type": "text", "text": "Describe this image in one sentence."},
        ],
    }
]
inputs = processor.apply_chat_template(
    messages, tokenize=True, add_generation_prompt=True, return_dict=True, return_tensors="pt"
).to(model.device)

generated_ids = model.generate(**inputs, max_new_tokens=30, do_sample=False)
generated_ids = generated_ids[:, inputs.input_ids.shape[1] :]
print(processor.batch_decode(generated_ids, skip_special_tokens=True)[0])
```

```
(env_main) ➜  qwen3_vl_moe_tp_plan git:(qwen3_vl_moe_tp_plan) ✗ torchrun --nproc-per-node 4 repro.py
W0827 12:47:39.173000 4059001 torch/distributed/run.py:851] 
W0827 12:47:39.173000 4059001 torch/distributed/run.py:851] *****************************************
W0827 12:47:39.173000 4059001 torch/distributed/run.py:851] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
W0827 12:47:39.173000 4059001 torch/distributed/run.py:851] *****************************************
Skipping import of cpp extensions due to incompatible torch version 2.11.0+cu128 for torchao version 0.16.0             Please see https://github.com/pytorch/ao/issues/2919 for more info
Skipping import of cpp extensions due to incompatible torch version 2.11.0+cu128 for torchao version 0.16.0             Please see https://github.com/pytorch/ao/issues/2919 for more info
Skipping import of cpp extensions due to incompatible torch version 2.11.0+cu128 for torchao version 0.16.0             Please see https://github.com/pytorch/ao/issues/2919 for more info
Skipping import of cpp extensions due to incompatible torch version 2.11.0+cu128 for torchao version 0.16.0             Please see https://github.com/pytorch/ao/issues/2919 for more info
Loading weights: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 882/882 [00:04<00:00, 185.92it/s]
Loading weights: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 882/882 [00:04<00:00, 185.29it/s]
Loading weights: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 882/882 [00:04<00:00, 185.23it/s]
Loading weights: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 882/882 [00:04<00:00, 184.67it/s]
git16.363018752
 p16.363018752
16.363018752
us16.363018752
A fluffy Pallas's cat walks through the snow in front of a backdrop of birch trees.
A fluffy Pallas's cat walks through the snow in front of a backdrop of birch trees.
A fluffy Pallas's cat walks through the snow in front of a backdrop of birch trees.
A fluffy Pallas's cat walks through the snow in front of a backdrop of birch trees.
```

### CODE REVIEW

**Comment by 3outeille** [comment]
run-slow: qwen3_vl_moe

**Comment by Noxtimo** [comment]
Thanks for investigating the root cause and resolving the checkpoint transposition issue in the loader engine @3outeille!

The target-to-source dimension mapping (`get_source_dim_mapping`) in `DtensorShardOperation` makes complete sense for avoiding full checkpoint materialization during shard-on-read.

One small note: I noticed the model-level unit test `test_tp_plan` from #48354 in `tests/models/qwen3_vl_moe/test_modeling_qwen3_vl_moe.py` was omitted here. Adding that test alongside the engine tests in `test_core_model_loading.py` would help prevent regressions to `Qwen3VLMoe`'s TP plan in the future.

**Comment by Noxtimo** [comment]
Also, since the `Qwen3-VL-MoE` TP & FSDP plan rules and definitions (`configuration_qwen3_vl_moe.py`, `modeling_qwen3_vl_moe.py`, `modular_qwen3_vl_moe.py`) originated from #48354, could you please add co-authorship attribution on the squash merge commit when this lands?

```
Co-authored-by: Noxtimo <139155109+Noxtimo@users.noreply.github.com>
```

Appreciate it!

**Comment by 3outeille** [comment]
sure will add co-author !

File: src/transformers/models/qwen3_vl_moe/modular_qwen3_vl_moe.py [hunk 1]
```diff
@@ -337,6 +342,9 @@ class Qwen3VLMoeCausalLMOutputWithPast(Qwen3VLCausalLMOutputWithPast):
 
 
 class Qwen3VLMoeForConditionalGeneration(Qwen3VLForConditionalGeneration):
+    _tp_plan = {"lm_head": "colwise_gather_output"}
+    _fsdp_plan = {"lm_head": "keep_full_weight"}
```
Comment (ArthurZucker) [inline src/transformers/models/qwen3_vl_moe/modular_qwen3_vl_moe.py]: as I said in general these should be default for conditional generation (otherwise its "kinda" unrelated code for everyone)!

**Comment by qgallouedec** [comment]
Is this PR ready top be merged?

**Comment by Noxtimo** [comment]
Yes, from my side this is good to merge! The loader dimension mapping handles the transposed checkpoint weights cleanly, the model TP rules are set, and the test suite is green. Just a friendly reminder to please preserve the co-authorship attribution on the squash merge commit when landing:

```
Co-authored-by: Noxtimo <139155109+Noxtimo@users.noreply.github.com>
```

Thanks @3outeille and @qgallouedec!

**Comment by VI-Arthur** [comment]
Nope not yet gimme a bit of time! On it
